trigger/src/lib.rs

```diff
@@ -0,0 +1,290 @@
+//! Smartcontract which creates new NFT for every user
+#![no_std]
+
+extern crate alloc;
+#[cfg(not(test))]
+extern crate panic_halt;
+
+use core::ops::ControlFlow;
+
+use alloc::borrow::ToOwned;
+use alloc::collections::btree_map::BTreeMap;
+use alloc::collections::btree_set::BTreeSet;
+use alloc::vec::Vec;
+
+use dlmalloc::GlobalDlmalloc;
+use eyre::{Context as _, OptionExt as _, Result, bail, eyre};
+use iroha_crypto::SignatureOf;
+use iroha_trigger::data_model::block::BlockHeader;
+use iroha_trigger::log::*;
+use iroha_trigger::prelude::*;
+use serde::{Deserialize, Serialize};
+
+#[global_allocator]
+static ALLOC: GlobalDlmalloc = GlobalDlmalloc;
+
+#[derive(Deserialize)]
+struct Config {
+    /// Operation mode of this trigger - hub or domestic.
+    ///
+    /// Alters the behaviour of how transfers are made upon verification.
+    mode: OperationMode,
+    /// Storage with admin-only write access. Contains [`ChainSnapshot`]\(s).
+    ///
+    /// The trigger is deployed separately for each connected chain. Thus, there is
+    /// only one trigger on each domestic chain (domestic-hub), and many triggers
+    /// on the hub chain (hub-\*domestic). Therefore, on the hub chain, we use
+    /// a single admin storage to store multiple chain snapshots. We deploy a separate
+    /// trigger for each chain working with its own entry in the admin store.
+    admin_store: NftId,
```

Comment:
I think the key-value store doesn’t need to be implemented as an NFT; in the future, account metadata or similar could be used instead.